pkg/controlplane/controller/kubernetesservice/controller.go

```diff
@@ -121,72 +81,11 @@ func (c *Controller) Start() {
 		klog.Errorf("Error removing old endpoints from kubernetes service: %v", err)
 	}
 
-	repairNodePorts := portallocatorcontroller.NewRepair(c.ServiceNodePortInterval, c.Client.CoreV1(), c.Client.EventsV1(), c.ServiceNodePortRange, c.ServiceNodePortRegistry)
-
-	// We start both repairClusterIPs and repairNodePorts to ensure repair
-	// loops of ClusterIPs and NodePorts.
-	// We run both repair loops using RunUntil public interface.
-	// However, we want to fail liveness/readiness until the first
-	// successful repair loop, so we basically pass appropriate
-	// callbacks to RunUtil methods.
-	// Additionally, we ensure that we don't wait for it for longer
-	// than 1 minute for backward compatibility of failing the whole
-	// apiserver if we can't repair them.
-	wg := sync.WaitGroup{}
-	wg.Add(1)
-
-	runRepairNodePorts := func(stopCh chan struct{}) {
-		repairNodePorts.RunUntil(wg.Done, stopCh)
-	}
-
-	wg.Add(1)
-	var runRepairClusterIPs func(stopCh chan struct{})
-	if !utilfeature.DefaultFeatureGate.Enabled(features.MultiCIDRServiceAllocator) {
-		repairClusterIPs := servicecontroller.NewRepair(c.ServiceClusterIPInterval,
-			c.Client.CoreV1(),
-			c.Client.EventsV1(),
-			&c.ServiceClusterIPRange,
-			c.ServiceClusterIPRegistry,
-			&c.SecondaryServiceClusterIPRange,
-			c.SecondaryServiceClusterIPRegistry)
-		runRepairClusterIPs = func(stopCh chan struct{}) {
-			repairClusterIPs.RunUntil(wg.Done, stopCh)
-		}
-	} else {
-		repairClusterIPs := servicecontroller.NewRepairIPAddress(c.ServiceClusterIPInterval,
-			c.Client,
-			&c.ServiceClusterIPRange,
-			&c.SecondaryServiceClusterIPRange,
-			c.Informers.Core().V1().Services(),
-			c.Informers.Networking().V1alpha1().IPAddresses(),
-		)
-		runRepairClusterIPs = func(stopCh chan struct{}) {
-			repairClusterIPs.RunUntil(wg.Done, stopCh)
-		}
-	}
-	c.runner = async.NewRunner(c.RunKubernetesService, runRepairClusterIPs, runRepairNodePorts)
-	c.runner.Start()
-
-	// For backward compatibility, we ensure that if we never are able
-	// to repair clusterIPs and/or nodeports, we not only fail the liveness
-	// and/or readiness, but also explicitly fail.
-	done := make(chan struct{})
-	go func() {
-		defer close(done)
-		wg.Wait()
-	}()
-	select {
-	case <-done:
-	case <-time.After(time.Minute):
-		klog.Fatalf("Unable to perform initial IP and Port allocation check")
-	}
+	go c.Run(stopCh)
 }
 
 // Stop cleans up this API Servers endpoint reconciliation leases so another master can take over more quickly.
 func (c *Controller) Stop() {
-	if c.runner != nil {
-		c.runner.Stop()
```

Comment:
the thing is that the Stop() here need to be synchronized with the stopCh of the Start() method, or this can race creating and deleting the Service and its endpoints, thing is Stop() has to run only after Run() has stopped

Reply:
Won't this data-race on `c.runner`? Looks all very fragile.

Reply:
Double checked the runner semantics: it does not wait for the running process to terminate. It only closes the channel. So we need something equivalent. But it also means that the old code had a race here.

Reply:
Readded stop logic. I don't think we need it in our setup with hooks. But for the semantics of the controller alone it makes sense.